### PR TITLE
GH-43202: [C++][Compute] Detect and explicit error for offset overflow in row table

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

There are two places for the offset in a row table to possibly overflow: 1) Encoding columns into a row table; 2) Appending to a row table from another row table.

They are particularly bad because they can cause silent wrong result for the computation.

### What changes are included in this PR?

Detect such overflows in the aforementioned places and report an explicit error when an overflow is detected.

### Are these changes tested?

UT included.

### Are there any user-facing changes?

User code that used to be fake passing could see an explicit error after this change.

**This PR contains a "Critical Fix".**
* GitHub Issue: #43202

### CODE REVIEW

**Comment by zanmato1984** [comment]
Hi @pitrou @westonpace , would you like to take a look? Thanks.

File: cpp/src/arrow/compute/row/encode_internal.cc [hunk 1]
```diff
@@ -160,8 +161,8 @@ Status RowTableEncoder::EncodeSelected(RowTableImpl* rows, uint32_t num_selected
                         /*num_extra_bytes_to_append=*/static_cast<uint32_t>(0)));
   // Then populate the offsets of the var-length columns, which will be used as the target
   // size of the var-length buffers resizing below.
-  EncoderOffsets::GetRowOffsetsSelected(rows, batch_varbinary_cols_, num_selected,
-                                        selection);
+  RETURN_NOT_OK(EncoderOffsets::GetRowOffsetsSelected(rows, batch_varbinary_cols_,
```
Comment (zanmato1984) [inline cpp/src/arrow/compute/row/encode_internal.cc:164]: This is what I've been always curious about as my question in #43209. If anyone sees this and comment on #43209, I'll be much appreciated.

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/arrow/compute/row/row_test.cc [hunk 2]
```diff
@@ -125,5 +125,134 @@ TEST(RowTableMemoryConsumption, Encode) {
...
+  row_encoder.Init(column_metadatas, row_alignment, var_length_alignment);
+
+  // The rows to encode.
+  std::vector<uint16_t> row_ids(num_rows, 0);
+
+  // Encoding 7 rows should be fine.
+  {
+    row_encoder.PrepareEncodeSelected(0, num_rows - 1, columns);
+    ASSERT_OK(row_encoder.EncodeSelected(&row_table, static_cast<uint32_t>(num_rows - 1),
+                                         row_ids.data()));
```
[the remark this answers is not in the document]
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/row_test.cc:172]: Thanks for looking!

Every time an encoder encodes into a row table, the row table gets effectively cleared. So the memory is only hold til the next `EncodeSelected` call. Clearing it here doesn't make much difference I guess(?) - the code in between is relatively trivial.

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/row/row_test.cc:172]: Ah, ok, thank you

**Review by pitrou — APPROVED** [review]

**Pull request merged**